### PR TITLE
Kinesis ehfo

### PR DESCRIPTION
**Pull request by addisonj**

### CODE REVIEW

File: pulsar-io/kinesis/src/main/java/org/apache/pulsar/io/kinesis/AwsCredentialProviderPlugin.java [hunk 1]
```diff
@@ -47,4 +49,27 @@
      */
     AWSCredentialsProvider getCredentialProvider();
 
+    /**
+     * Returns a V2 credential provider for use with the v2 SDK.
+     *
+     * Defaults to an implementation that pulls credentials from a v1 provider
+     */
+    default software.amazon.awssdk.auth.credentials.AwsCredentialsProvider getV2CredentialsProvider() {
+        // make a small wrapper to forward requests to v1, this alows
```
Comment (feathj) [inline pulsar-io/kinesis/src/main/java/org/apache/pulsar/io/kinesis/AwsCredentialProviderPlugin.java]: Incomplete thought here

File: pulsar-io/kinesis/src/main/java/org/apache/pulsar/io/kinesis/KinesisRecord.java [hunk 2]
```diff
@@ -40,21 +41,30 @@
...
             try {
-                s = decoder.decode(record.getData()).toString();
+                s = decoder.decode(record.data()).toString();
             } catch (CharacterCodingException e) {
                // Ignore 
             }
             this.value = (s != null) ? s.getBytes() : null;
+        } else if (encType == EncryptionType.KMS) {
+            // use the raw encrypted value, let them handle it downstream
+            // TODO: support decoding KMS data here... should be fairly simple
```
Comment (feathj) [inline pulsar-io/kinesis/src/main/java/org/apache/pulsar/io/kinesis/KinesisRecord.java:66]: Do we want to document this as a known limitation for now? Not sure where that would even make sense to call out.

File: pulsar-io/kinesis/src/main/java/org/apache/pulsar/io/kinesis/KinesisRecordProcessor.java [hunk 3]
```diff
@@ -112,4 +79,45 @@ private void checkpoint(IRecordProcessorCheckpointer checkpointer) {
...
 
+    @Override
+    public void initialize(InitializationInput initializationInput) {
+        kinesisShardId = initializationInput.shardId();
+    }
+
+    @Override
+    public void processRecords(ProcessRecordsInput processRecordsInput) {
+
+        log.info("Processing " + processRecordsInput.records().size() + " records from " + kinesisShardId);
```
Comment (feathj) [inline pulsar-io/kinesis/src/main/java/org/apache/pulsar/io/kinesis/KinesisRecordProcessor.java:90]: does "records" ever return null here? Would hate to throw a null pointer exception in a log statement. Do we need to do a null block?

File: pulsar-io/kinesis/src/main/java/org/apache/pulsar/io/kinesis/KinesisRecordProcessor.java [hunk 4]
```diff
@@ -112,4 +79,45 @@ private void checkpoint(IRecordProcessorCheckpointer checkpointer) {
...
+        // Checkpoint once every checkpoint interval.
+        if (System.nanoTime() > nextCheckpointTimeInNanos) {
+            checkpoint(processRecordsInput.checkpointer());
+            nextCheckpointTimeInNanos = System.nanoTime() + checkpointInterval;
+        }
+    }
+
+    @Override
+    public void leaseLost(LeaseLostInput leaseLostInput) {
+        log.info("lease lost, will terminate soon");
```
Comment (feathj) [inline pulsar-io/kinesis/src/main/java/org/apache/pulsar/io/kinesis/KinesisRecordProcessor.java:109]: Seems strange to not do anything with the leaseLostInput object here

File: pulsar-io/kinesis/src/main/java/org/apache/pulsar/io/kinesis/KinesisSource.java [hunk 5]
```diff
@@ -80,31 +85,58 @@ public void open(Map<String, Object> config, SourceContext sourceContext) throws
...
+        RetrievalConfig retrievalConfig = configsBuilder.retrievalConfig();
+        if (!kinesisSourceConfig.isUseEnhancedFanOut()) {
+            retrievalConfig.retrievalSpecificConfig(
+                    new PollingConfig(kinesisSourceConfig.getAwsKinesisStreamName(),
+                                      kClient));
+        }
+
+        retrievalConfig.initialPositionInStreamExtended(kinesisSourceConfig.getStreamStartPosition());
+
+        scheduler = new Scheduler(
```
Comment (feathj) [inline pulsar-io/kinesis/src/main/java/org/apache/pulsar/io/kinesis/KinesisSource.java:111]: I don't remember seeing any of this scheduler / threading stuff in their code. They must not have ever tested it.

**Review by feathj — CHANGES_REQUESTED** [review]
Just a couple of comments / suggestions

File: pulsar-io/kinesis/src/main/java/org/apache/pulsar/io/kinesis/KinesisSource.java (hunk 5 above)
Comment (addisonj) [reply pulsar-io/kinesis/src/main/java/org/apache/pulsar/io/kinesis/KinesisSource.java:111]: the scheduler is a new thing introduced in v2 of the KCL, but there was an analogous thing called a "worker". They failed to start the worker in another thread which was why it didn't work :P

File: pulsar-io/kinesis/src/main/java/org/apache/pulsar/io/kinesis/KinesisRecordProcessor.java (hunk 4 above)
Comment (addisonj) [reply pulsar-io/kinesis/src/main/java/org/apache/pulsar/io/kinesis/KinesisRecordProcessor.java:109]: the KCL docs seem to indicate we will exit if we get here, hence the log statement

File: pulsar-io/kinesis/src/main/java/org/apache/pulsar/io/kinesis/KinesisRecordProcessor.java (hunk 3 above)
Comment (addisonj) [reply pulsar-io/kinesis/src/main/java/org/apache/pulsar/io/kinesis/KinesisRecordProcessor.java:90]: it should not, as the kcl v2 with enhanced fan-out is all pushed based, so it shouldn't happen

**Comment by feathj** [comment]
Approved. :shipit:

**Pull request merged**